flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -999,6 +999,93 @@ class Table(
     new OverWindowedTable(this, overWindows.toArray)
   }
 
+  /**
+    * Performs a map operation with an user-defined function. The output will be flattened
+    * if the output type is composite type.
+    *
+    * Scala Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     def eval(str: String): Row = {
+    *       Row.of(str, str)
+    *     }
+    *
+    *     def getResultType(signature: Array[Class[_]]): TypeInformation[_] =
+    *       Types.ROW(Types.STRING, Types.STRING)
+    *   }
+    *
+    *   val func = new MyMapFunction()
+    *   table.map(func('c))
+    * }}}
+    *
+    * Java Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     public Row eval(String str) {
+    *       return Row.of(str, str);
+    *     }
+    *
+    *     public TypeInformation getResultType(Class[] signature) {
+    *       return new RowTypeInfo(BasicTypeInfo.STRING_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO);
+    *     }
+    *   }
+    *
+    *   ScalarFunction func = new MyMapFunction();
+    *   tableEnv.registerFunction("func", func);
+    *   table.map("func(c)");
+    * }}}
+    */
+  def map(mapFunction: Expression): Table = {
```

Comment:
Can this be a ScalarFunctionCall? In this way, we even don't have to validate input types.

Reply:
`Expression` makes sure that it will not break backward compatible if we want to support `Expression*` in the future.

Reply:
It won't break backward compatible if we change `map(mapFunction: ScalarFunctionCall)` to `map(fields: Expression*)` since `ScalarFunctionCall` is also a `Expression`.

Reply:
You're right. +1

Reply:
I found that we need also support the built-in functions such "concat" which are parsed as `Call` instead of `ScalarFunctionCall` and so I prefer to use the parameter type of `Expression`.

Reply:
It is good if `map` can also support buildin scalar functions. Let's do it with `Expression`.